From: Jacks Diao <kdha200501@gmail.com>
Subject: [PATCH] jacks-customizations: one focus at a time
---
Index: src/filewidgets/kdiroperator.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/filewidgets/kdiroperator.cpp b/src/filewidgets/kdiroperator.cpp
--- a/src/filewidgets/kdiroperator.cpp	(revision 412d25b9cb60da4b4ec3f2ceebaea1ad4fbdb869)
+++ b/src/filewidgets/kdiroperator.cpp	(revision 566843bb86541113fb169ddb11d2d36df7b6ac08)
@@ -1758,6 +1758,9 @@
     });
     connect(d->m_dirLister, qOverload<>(&KCoreDirLister::completed), this, [this]() {
         d->slotIOFinished();
+        if(view()) {
+            view()->setFocus();
+        }
     });
     connect(d->m_dirLister, qOverload<>(&KCoreDirLister::canceled), this, [this]() {
         d->slotCanceled();
Index: src/filewidgets/kfilewidget.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/filewidgets/kfilewidget.cpp b/src/filewidgets/kfilewidget.cpp
--- a/src/filewidgets/kfilewidget.cpp	(revision 412d25b9cb60da4b4ec3f2ceebaea1ad4fbdb869)
+++ b/src/filewidgets/kfilewidget.cpp	(revision 566843bb86541113fb169ddb11d2d36df7b6ac08)
@@ -2112,6 +2112,7 @@
         m_urlNavigator->setUrlEditable(true);
         m_urlNavigator->setFocus();
         lineEdit->selectAll();
+        m_ops->view()->clearSelection();
     }
 }
 
Index: src/filewidgets/kurlnavigator.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/filewidgets/kurlnavigator.cpp b/src/filewidgets/kurlnavigator.cpp
--- a/src/filewidgets/kurlnavigator.cpp	(revision 412d25b9cb60da4b4ec3f2ceebaea1ad4fbdb869)
+++ b/src/filewidgets/kurlnavigator.cpp	(revision 566843bb86541113fb169ddb11d2d36df7b6ac08)
@@ -434,6 +434,10 @@
         };
         QMetaObject::invokeMethod(q, switchModeFunc, Qt::QueuedConnection);
     }
+
+    if(m_editable) {
+        switchView();
+    }
 }
 
 void KUrlNavigatorPrivate::slotSchemeChanged(const QString &scheme)
@@ -1136,6 +1140,8 @@
 {
     if (isUrlEditable() && (event->key() == Qt::Key_Escape)) {
         setUrlEditable(false);
+        d->m_pathBox->setUrl(d->m_coreUrlNavigator->currentLocationUrl());
+        Q_EMIT returnPressed();
     } else {
         QWidget::keyPressEvent(event);
     }
@@ -1209,6 +1215,15 @@
     // Avoid the "Properties" action from triggering instead of new tab.
     case QEvent::ShortcutOverride: {
         auto *keyEvent = static_cast<QKeyEvent *>(event);
+
+        // in case the return key is mapped to F2, but that would be the wrong thing to do because,
+        // macOS does not use the enter key to initiate a rename in the file dialog
+        // if (keyEvent->key() == Qt::Key_F2) {
+        //     Q_EMIT d->m_pathBox->returnPressed(d->m_pathBox->currentText());
+        //     event->ignore();
+        //     return false;
+        // }
+
         if ((keyEvent->key() == Qt::Key_Enter || keyEvent->key() == Qt::Key_Return)
             && (keyEvent->modifiers() & Qt::AltModifier || keyEvent->modifiers() & Qt::ShiftModifier)) {
             event->accept();
