From: Jacks Diao <kdha200501@gmail.com>
Subject: [PATCH] jacks-customizations: customize window decoration
---
Index: src/plugins/kdecorations/aurorae/src/qml/aurorae.qml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/plugins/kdecorations/aurorae/src/qml/aurorae.qml b/src/plugins/kdecorations/aurorae/src/qml/aurorae.qml
--- a/src/plugins/kdecorations/aurorae/src/qml/aurorae.qml	(revision 1a24f02af8383f470c1cd4aa676c11d5d848f271)
+++ b/src/plugins/kdecorations/aurorae/src/qml/aurorae.qml	(revision 72866971db5b8d7e889e45c149a2948a8a7ce1ed)
@@ -13,14 +13,14 @@
     property alias decorationMask: maskItem.mask
     property alias supportsMask: backgroundSvg.supportsMask
     Component.onCompleted: {
-        borders.left   = Qt.binding(function() { return Math.max(0, auroraeTheme.borderLeft);});
-        borders.right  = Qt.binding(function() { return Math.max(0, auroraeTheme.borderRight);});
-        borders.top    = Qt.binding(function() { return Math.max(0, auroraeTheme.borderTop);});
+        borders.left   = Qt.binding(function() { return Math.max(0, auroraeTheme.borderLeft - 1);});
+        borders.right  = Qt.binding(function() { return Math.max(0, auroraeTheme.borderRight + decoration.client.maximized ? 3 : 0);});
+        borders.top    = Qt.binding(function() { return Math.max(0, auroraeTheme.borderTop - 7);});
         borders.bottom = Qt.binding(function() { return Math.max(0, auroraeTheme.borderBottom);});
-        maximizedBorders.left   = Qt.binding(function() { return Math.max(0, auroraeTheme.borderLeftMaximized);});
-        maximizedBorders.right  = Qt.binding(function() { return Math.max(0, auroraeTheme.borderRightMaximized);});
+        maximizedBorders.left   = Qt.binding(function() { return Math.max(0, auroraeTheme.borderLeftMaximized - 1);});
+        maximizedBorders.right  = Qt.binding(function() { return Math.max(0, auroraeTheme.borderRightMaximized + decoration.client.maximized ? 3 : 0);});
         maximizedBorders.bottom = Qt.binding(function() { return Math.max(0, auroraeTheme.borderBottomMaximized);});
-        maximizedBorders.top    = Qt.binding(function() { return Math.max(0, auroraeTheme.borderTopMaximized);});
+        maximizedBorders.top    = Qt.binding(function() { return Math.max(0, auroraeTheme.borderTopMaximized - 7);});
         padding.left   = auroraeTheme.paddingLeft;
         padding.right  = auroraeTheme.paddingRight;
         padding.bottom = auroraeTheme.paddingBottom;
@@ -58,7 +58,7 @@
         imagePath: backgroundSvg.imagePath
         prefix: "decoration"
         opacity: shown ? 1 : 0
-        enabledBorders: decoration.client.maximized ? KSvg.FrameSvg.NoBorder : KSvg.FrameSvg.TopBorder | KSvg.FrameSvg.BottomBorder | KSvg.FrameSvg.LeftBorder | KSvg.FrameSvg.RightBorder
+        enabledBorders: KSvg.FrameSvg.TopBorder | KSvg.FrameSvg.BottomBorder | KSvg.FrameSvg.LeftBorder | KSvg.FrameSvg.RightBorder
         Behavior on opacity {
             enabled: root.animate
             NumberAnimation {
@@ -125,75 +125,173 @@
             }
         }
     }
-    AuroraeButtonGroup {
+//    AuroraeButtonGroup {
+//        id: leftButtonGroup
+//        buttons: options.titleButtonsLeft
+//        width: childrenRect.width
+//        animate: root.animate
+//        anchors {
+//            left: root.left
+//            leftMargin: decoration.client.maximized ? auroraeTheme.titleEdgeLeftMaximized : (auroraeTheme.titleEdgeLeft + root.padding.left)
+//        }
+//    }
+//    AuroraeButtonGroup {
+//        id: rightButtonGroup
+//        buttons: options.titleButtonsRight
+//        width: childrenRect.width
+//        animate: root.animate
+//        anchors {
+//            right: root.right
+//            rightMargin: decoration.client.maximized ? auroraeTheme.titleEdgeRightMaximized : (auroraeTheme.titleEdgeRight + root.padding.right)
+//        }
+//    }
+//    Text {
+//        id: caption
+//        text: decoration.client.caption
+//        textFormat: Text.PlainText
+//        horizontalAlignment: auroraeTheme.horizontalAlignment
+//        verticalAlignment: auroraeTheme.verticalAlignment
+//        elide: Text.ElideRight
+//        height: Math.max(auroraeTheme.titleHeight, auroraeTheme.buttonHeight * auroraeTheme.buttonSizeFactor)
+//        color: decoration.client.active ? auroraeTheme.activeTextColor : auroraeTheme.inactiveTextColor
+//        font: options.titleFont
+//        renderType: Text.NativeRendering
+//        anchors {
+//            left: leftButtonGroup.right
+//            right: rightButtonGroup.left
+//            top: root.top
+//            topMargin: decoration.client.maximized ? auroraeTheme.titleEdgeTopMaximized : (auroraeTheme.titleEdgeTop + root.padding.top)
+//            leftMargin: auroraeTheme.titleBorderLeft
+//            rightMargin: auroraeTheme.titleBorderRight
+//        }
+//        Behavior on color {
+//            enabled: root.animate
+//            ColorAnimation {
+//                duration: auroraeTheme.animationTime
+//            }
+//        }
+//    }
+    KSvg.FrameSvgItem {
+        id: innerBorder
+        anchors {
+            fill: parent
+            leftMargin: parent.padding.left + parent.borders.left - margins.left
+            rightMargin: parent.padding.right + parent.borders.right - margins.right
+            topMargin: parent.padding.top + parent.borders.top - margins.top
+            bottomMargin: parent.padding.bottom + parent.borders.bottom - margins.bottom
+        }
+        visible: parent.borders.left > fixedMargins.left
+            && parent.borders.right > fixedMargins.right
+            && parent.borders.top > fixedMargins.top
+            && parent.borders.bottom > fixedMargins.bottom
+
+        imagePath: backgroundSvg.imagePath
+        prefix: "innerborder"
+        opacity: (decoration.client.active && !decoration.client.maximized && backgroundSvg.supportsInnerBorder) ? 1 : 0
+        Behavior on opacity {
+            enabled: root.animate
+            NumberAnimation {
+                duration: auroraeTheme.animationTime
+            }
+        }
+    }
+
+    Rectangle {
         id: leftButtonGroup
-        buttons: options.titleButtonsLeft
-        width: childrenRect.width
-        animate: root.animate
+        height: childrenRect.height
+        width: childrenRect.width + 6
+        anchors {
+            top: root.top
+            left: root.left
+            topMargin: 2
+            rightMargin: 0
+            bottomMargin: 0
+            leftMargin: 2
+        }
+        color: decoration.client.active ? "#CCCCCC" : "#DDDDDD"
+        AuroraeButtonGroup {
+            id: leftButtonGroupCta
+            buttons: options.titleButtonsLeft
+            height: childrenRect.height
+            width: childrenRect.width
+            anchors {
+                top: parent.top
+                horizontalCenter: parent.horizontalCenter
+                topMargin: 1
+                rightMargin: 1
+                bottomMargin: 1
+                leftMargin: 1
+            }
+            animate: root.animate
+        }
+    }
+    Rectangle {
+        id: rightButtonGroup
+        height: childrenRect.height
+        width: childrenRect.width + 6
         anchors {
-            left: root.left
-            leftMargin: decoration.client.maximized ? auroraeTheme.titleEdgeLeftMaximized : (auroraeTheme.titleEdgeLeft + root.padding.left)
+            top: root.top
+            right: root.right
+            topMargin: 2
+            rightMargin: 4
+            bottomMargin: 0
+            leftMargin: 0
         }
-    }
-    AuroraeButtonGroup {
-        id: rightButtonGroup
-        buttons: options.titleButtonsRight
-        width: childrenRect.width
-        animate: root.animate
+        color: decoration.client.active ? "#CCCCCC" : "#DDDDDD"
+        AuroraeButtonGroup {
+              id: rightButtonGroupCta
+              buttons: options.titleButtonsRight
+              height: childrenRect.height
+              width: childrenRect.width
+              anchors {
+                  top: parent.top
+                  horizontalCenter: parent.horizontalCenter
+                  topMargin: 1
+                  rightMargin: 1
+                  bottomMargin: 1
+                  leftMargin: 1
+              }
+              animate: root.animate
+          }
+    }
+    Rectangle {
+        id: caption
+        height: 15
+        width: captionMetrics.width + 15
         anchors {
-            right: root.right
-            rightMargin: decoration.client.maximized ? auroraeTheme.titleEdgeRightMaximized : (auroraeTheme.titleEdgeRight + root.padding.right)
+            horizontalCenter: parent.horizontalCenter
+            //left: leftButtonGroup.right
+            //right: rightButtonGroup.left
+            top: root.top
+            topMargin: auroraeTheme.titleEdgeTop + root.padding.top + 2
+            leftMargin: auroraeTheme.titleBorderLeft
+            rightMargin: auroraeTheme.titleBorderRight
         }
-    }
-    Text {
-        id: caption
-        text: decoration.client.caption
-        textFormat: Text.PlainText
-        horizontalAlignment: auroraeTheme.horizontalAlignment
-        verticalAlignment: auroraeTheme.verticalAlignment
-        elide: Text.ElideRight
-        height: Math.max(auroraeTheme.titleHeight, auroraeTheme.buttonHeight * auroraeTheme.buttonSizeFactor)
-        color: decoration.client.active ? auroraeTheme.activeTextColor : auroraeTheme.inactiveTextColor
-        font: options.titleFont
-        renderType: Text.NativeRendering
-        anchors {
-            left: leftButtonGroup.right
-            right: rightButtonGroup.left
-            top: root.top
-            topMargin: decoration.client.maximized ? auroraeTheme.titleEdgeTopMaximized : (auroraeTheme.titleEdgeTop + root.padding.top)
-            leftMargin: auroraeTheme.titleBorderLeft
-            rightMargin: auroraeTheme.titleBorderRight
-        }
-        Behavior on color {
-            enabled: root.animate
-            ColorAnimation {
-                duration: auroraeTheme.animationTime
-            }
-        }
-    }
-    KSvg.FrameSvgItem {
-        id: innerBorder
-        anchors {
-            fill: parent
-            leftMargin: parent.padding.left + parent.borders.left - margins.left
-            rightMargin: parent.padding.right + parent.borders.right - margins.right
-            topMargin: parent.padding.top + parent.borders.top - margins.top
-            bottomMargin: parent.padding.bottom + parent.borders.bottom - margins.bottom
-        }
-        visible: parent.borders.left > fixedMargins.left
-            && parent.borders.right > fixedMargins.right
-            && parent.borders.top > fixedMargins.top
-            && parent.borders.bottom > fixedMargins.bottom
-
-        imagePath: backgroundSvg.imagePath
-        prefix: "innerborder"
-        opacity: (decoration.client.active && !decoration.client.maximized && backgroundSvg.supportsInnerBorder) ? 1 : 0
-        Behavior on opacity {
-            enabled: root.animate
-            NumberAnimation {
-                duration: auroraeTheme.animationTime
+        color: decoration.client.active ? "#CCCCCC" : "#DDDDDD"
+        Text {
+            id: captionText
+            text: decoration.client.caption
+            textFormat: Text.PlainText
+            horizontalAlignment: auroraeTheme.horizontalAlignment
+            verticalAlignment: auroraeTheme.verticalAlignment
+            elide: Text.ElideRight
+            color: decoration.client.active ? auroraeTheme.activeTextColor : auroraeTheme.inactiveTextColor
+            //font: options.titleFont
+            font.pixelSize: 12
+            renderType: Text.NativeRendering
+            anchors.centerIn: parent
+            Behavior on color {
+                enabled: root.animate
+                ColorAnimation {
+                    duration: auroraeTheme.animationTime
+                }
             }
         }
+        TextMetrics {
+            id: captionMetrics
+            text: decoration.client.caption
+            font: options.titleFont
+        }
     }
     KSvg.FrameSvgItem {
         id: innerBorderInactive
