From: Jacks Diao <kdha200501@gmail.com>
Subject: [PATCH] jacks-customizations: bring back shading
---
Index: src/decorations/decoratedwindow.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/decorations/decoratedwindow.cpp b/src/decorations/decoratedwindow.cpp
--- a/src/decorations/decoratedwindow.cpp	(revision 9739e3174abaaa97a3008f6933af4642bc26b800)
+++ b/src/decorations/decoratedwindow.cpp	(revision 467bd2c1d82f46d0b2629411ef9dbec1f7d01f57)
@@ -170,7 +170,7 @@

 void DecoratedWindowImpl::requestMinimize()
 {
-    m_window->setMinimized(true);
+    delayedRequestToggleMaximization(Options::ShadeOp);
 }

 void DecoratedWindowImpl::requestClose()
Index: src/effect/globals.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/effect/globals.h b/src/effect/globals.h
--- a/src/effect/globals.h	(revision 9739e3174abaaa97a3008f6933af4642bc26b800)
+++ b/src/effect/globals.h	(revision 467bd2c1d82f46d0b2629411ef9dbec1f7d01f57)
@@ -59,6 +59,7 @@
     MaximizeHorizontal = 2, ///< The window is maximized horizontally.
     /// Equal to @p MaximizeVertical | @p MaximizeHorizontal
     MaximizeFull = MaximizeVertical | MaximizeHorizontal,
+    MaximizeShade = 1027,
 };
 Q_ENUM_NS(MaximizeMode)

Index: src/useractions.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/useractions.cpp b/src/useractions.cpp
--- a/src/useractions.cpp	(revision 9739e3174abaaa97a3008f6933af4642bc26b800)
+++ b/src/useractions.cpp	(revision 467bd2c1d82f46d0b2629411ef9dbec1f7d01f57)
@@ -897,7 +897,7 @@
     initShortcut("Window Minimize", i18n("Minimize Window"),
                  Qt::META | Qt::Key_PageDown, &Workspace::slotWindowMinimize);
     initShortcut("Window Shade", i18n("Shade Window"),
-                 0, &Workspace::slotWindowShade);
+                 Qt::CTRL | Qt::Key_M, &Workspace::slotWindowShade);
     initShortcut("Window Move", i18n("Move Window"),
                  0, &Workspace::slotWindowMove);
     initShortcut("Window Resize", i18n("Resize Window"),
@@ -1175,7 +1175,10 @@
         window->setMinimized(true);
         break;
     case Options::ShadeOp:
-        window->performMousePressCommand(Options::MouseShade, Cursors::self()->mouse()->pos());
+        window->maximize(window->maximizeMode() == MaximizeShade
+                                     ? MaximizeRestore
+                                     : MaximizeShade);
+        takeActivity(window, ActivityFocus | ActivityRaise);
         break;
     case Options::OnAllDesktopsOp:
         window->setOnAllDesktops(!window->isOnAllDesktops());
Index: src/wayland/xdgshell.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/wayland/xdgshell.h b/src/wayland/xdgshell.h
--- a/src/wayland/xdgshell.h	(revision 9739e3174abaaa97a3008f6933af4642bc26b800)
+++ b/src/wayland/xdgshell.h	(revision 467bd2c1d82f46d0b2629411ef9dbec1f7d01f57)
@@ -214,8 +214,8 @@

 public:
     enum State {
-        MaximizedHorizontal = 0x1,
-        MaximizedVertical = 0x2,
+        MaximizedHorizontal = 0x1,//                           000000000001 in practice, kwin treats this flag as size-changed-horizontally
+        MaximizedVertical = 0x2,//                             000000000010 in practice, kwin treats this flag as size-changed-vertically
         FullScreen = 0x4,
         Resizing = 0x8,
         Activated = 0x10,
@@ -224,7 +224,8 @@
         TiledRight = 0x80,
         TiledBottom = 0x100,
         Suspended = 0x200,
-        Maximized = MaximizedHorizontal | MaximizedVertical,
+        Maximized = MaximizedHorizontal | MaximizedVertical,// 000000000011
+        Shaded = 0x403,//                                      010000000011
     };
     Q_DECLARE_FLAGS(States, State)

Index: src/xdgshellwindow.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/xdgshellwindow.cpp b/src/xdgshellwindow.cpp
--- a/src/xdgshellwindow.cpp	(revision 9739e3174abaaa97a3008f6933af4642bc26b800)
+++ b/src/xdgshellwindow.cpp	(revision 467bd2c1d82f46d0b2629411ef9dbec1f7d01f57)
@@ -835,20 +835,17 @@
     scheduleConfigure();
 }

-void XdgToplevelWindow::doSetMaximized()
+void XdgToplevelWindow::doSetMaximized(MaximizeMode nextMaximizeMode, MaximizeMode currentMaximizeMode)
 {
-    if (requestedMaximizeMode() & MaximizeHorizontal) {
-        m_nextStates |= XdgToplevelInterface::State::MaximizedHorizontal;
-    } else {
-        m_nextStates &= ~XdgToplevelInterface::State::MaximizedHorizontal;
+    if (currentMaximizeMode == MaximizeShade && nextMaximizeMode == MaximizeFull) {
+        m_acknowledgedStates &= ~XdgToplevelInterface::State::Shaded;
     }

-    if (requestedMaximizeMode() & MaximizeVertical) {
-        m_nextStates |= XdgToplevelInterface::State::MaximizedVertical;
-    } else {
-        m_nextStates &= ~XdgToplevelInterface::State::MaximizedVertical;
+    if (currentMaximizeMode == MaximizeFull && nextMaximizeMode == MaximizeShade) {
+        m_acknowledgedStates &= ~XdgToplevelInterface::State::Maximized;
     }

+    setMaximizeModeToStates(nextMaximizeMode);
     scheduleConfigure();
 }

@@ -974,6 +971,43 @@
     return !isDeleted() && readyForPainting();
 }

+void XdgToplevelWindow::setMaximizeModeToStates(MaximizeMode nextMaximizeMode)
+{
+    if (nextMaximizeMode == MaximizeShade) {
+        m_nextStates |= XdgToplevelInterface::State::Shaded;
+        return;
+    }
+
+    m_nextStates &= ~XdgToplevelInterface::State::Shaded;
+
+    if (nextMaximizeMode & MaximizeHorizontal) {
+        m_nextStates |= XdgToplevelInterface::State::MaximizedHorizontal;
+    }
+
+    if (nextMaximizeMode & MaximizeVertical) {
+        m_nextStates |= XdgToplevelInterface::State::MaximizedVertical;
+    }
+}
+
+MaximizeMode XdgToplevelWindow::getMaximizeModeFromStates(const XdgToplevelInterface::States &states)
+{
+    if (states & (XdgToplevelInterface::State::Shaded & ~XdgToplevelInterface::State::Maximized)) {
+        return MaximizeShade;
+    }
+
+    MaximizeMode maximizeMode = MaximizeRestore;
+
+    if (states & XdgToplevelInterface::State::MaximizedHorizontal) {
+        maximizeMode = MaximizeMode(maximizeMode | MaximizeHorizontal);
+    }
+
+    if (states & XdgToplevelInterface::State::MaximizedVertical) {
+        maximizeMode = MaximizeMode(maximizeMode | MaximizeVertical);
+    }
+
+    return maximizeMode;
+}
+
 void XdgToplevelWindow::handleWindowTitleChanged()
 {
     setCaption(m_shellSurface->title());
@@ -1035,14 +1069,7 @@
     const XdgToplevelInterface::States delta = m_acknowledgedStates ^ states;

     if (delta & XdgToplevelInterface::State::Maximized) {
-        MaximizeMode maximizeMode = MaximizeRestore;
-        if (states & XdgToplevelInterface::State::MaximizedHorizontal) {
-            maximizeMode = MaximizeMode(maximizeMode | MaximizeHorizontal);
-        }
-        if (states & XdgToplevelInterface::State::MaximizedVertical) {
-            maximizeMode = MaximizeMode(maximizeMode | MaximizeVertical);
-        }
-        updateMaximizeMode(maximizeMode);
+        updateMaximizeMode(getMaximizeModeFromStates(states));
     }
     if (delta & XdgToplevelInterface::State::FullScreen) {
         updateFullScreenMode(states & XdgToplevelInterface::State::FullScreen);
@@ -1410,6 +1437,119 @@
     Q_EMIT maximizedChanged();
 }

+static bool changeMaximizeRecursion = false;
+void XdgToplevelWindow::animateMaximize(MaximizeMode nextMaximizeMode, MaximizeMode currentMaximizeMode)
+{
+    Q_EMIT maximizedAboutToChange(nextMaximizeMode);
+
+    if (!m_nextDecoration) {
+        return;
+    }
+
+    if (options->borderlessMaximizedWindows() && nextMaximizeMode == MaximizeFull) {
+        return;
+    }
+
+    changeMaximizeRecursion = true;
+    const auto c = m_nextDecoration->window();
+
+    if ((nextMaximizeMode & MaximizeVertical) != (currentMaximizeMode & MaximizeVertical)) {
+        Q_EMIT c->maximizedVerticallyChanged(nextMaximizeMode & MaximizeVertical);
+    }
+
+    if ((nextMaximizeMode & MaximizeHorizontal) != (currentMaximizeMode & MaximizeHorizontal)) {
+        Q_EMIT c->maximizedHorizontallyChanged(nextMaximizeMode & MaximizeHorizontal);
+    }
+
+    if ((nextMaximizeMode == MaximizeFull) != (currentMaximizeMode == MaximizeFull) || (nextMaximizeMode == MaximizeShade) != (currentMaximizeMode == MaximizeShade)) {
+        Q_EMIT c->maximizedChanged(nextMaximizeMode == MaximizeFull);
+    }
+
+    changeMaximizeRecursion = false;
+}
+
+QRectF XdgToplevelWindow::nextMaximizeHorizontalGeometry(const QRectF &clientArea, MaximizeMode nextMaximizeMode) {
+    QRectF nextGeometry = moveResizeGeometry();
+
+    // if the window is to be maximized
+    if (nextMaximizeMode & MaximizeHorizontal && nextMaximizeMode != MaximizeShade) {
+        // then take the full view port width
+        nextGeometry.setX(clientArea.x());
+        nextGeometry.setWidth(clientArea.width());
+        return nextGeometry;
+    }
+
+    QRectF previousGeometry = geometryRestore();
+
+    if (!previousGeometry.isValid()) {
+        nextGeometry.setX(clientArea.x());
+        nextGeometry.setWidth(0);
+        return nextGeometry;
+    }
+
+    // if the window is to be shaded or restored,
+    // then restore
+    nextGeometry.setX(previousGeometry.x());
+    nextGeometry.setWidth(previousGeometry.width());
+    return nextGeometry;
+}
+
+QRectF XdgToplevelWindow::nextMaximizeVerticalGeometry(const QRectF &clientArea, MaximizeMode nextMaximizeMode) {
+    QRectF nextGeometry = moveResizeGeometry();
+
+    // if the window is to be maximized
+    if (nextMaximizeMode & MaximizeVertical && nextMaximizeMode != MaximizeShade) {
+        // then take the full view port height
+        nextGeometry.setY(clientArea.y());
+        nextGeometry.setHeight(clientArea.height());
+        return nextGeometry;
+    }
+
+    QRectF previousGeometry = geometryRestore();
+
+    if (!previousGeometry.isValid()) {
+        nextGeometry.setY(clientArea.y());
+        nextGeometry.setHeight(0);
+        return nextGeometry;
+    }
+
+    // if the window is to be shaded
+    if (nextMaximizeMode & MaximizeVertical && nextMaximizeMode == MaximizeShade) {
+        // then shrink to the window title bar height
+        nextGeometry.setY(previousGeometry.y());
+        nextGeometry.setHeight(24);
+        return nextGeometry;
+    }
+
+    // if the window is to be restored,
+    // then restore
+    nextGeometry.setY(previousGeometry.y());
+    nextGeometry.setHeight(previousGeometry.height());
+    return nextGeometry;
+}
+
+void XdgToplevelWindow::backupCurrentMaximizeGeometry(MaximizeMode currentMaximizeMode)
+{
+    if (requestedQuickTileMode() != QuickTileMode(QuickTileFlag::None)) {
+        return;
+    }
+
+    QRectF currentGeometry = moveResizeGeometry();
+    QRectF backupGeometry = geometryRestore();
+
+    if (!(currentMaximizeMode & MaximizeVertical)) {
+        backupGeometry.setTop(currentGeometry.top());
+        backupGeometry.setBottom(currentGeometry.bottom());
+    }
+
+    if (!(currentMaximizeMode & MaximizeHorizontal)) {
+        backupGeometry.setLeft(currentGeometry.left());
+        backupGeometry.setRight(currentGeometry.right());
+    }
+
+    setGeometryRestore(backupGeometry);
+}
+
 void XdgToplevelWindow::updateFullScreenMode(bool set)
 {
     if (m_isFullScreen == set) {
@@ -1679,8 +1819,7 @@
     doSetFullScreen();
 }

-static bool changeMaximizeRecursion = false;
-void XdgToplevelWindow::maximize(MaximizeMode mode, const QRectF &restore)
+void XdgToplevelWindow::maximize(MaximizeMode nextMaximizeMode, const QRectF &restore)
 {
     if (changeMaximizeRecursion) {
         return;
@@ -1690,102 +1829,50 @@
         return;
     }

-    const QRectF clientArea = isElectricBorderMaximizing() ? workspace()->clientArea(MaximizeArea, this, interactiveMoveResizeAnchor()) : workspace()->clientArea(MaximizeArea, this, moveResizeOutput());
+    if (nextMaximizeMode != MaximizeShade) {
+        nextMaximizeMode = rules()->checkMaximize(nextMaximizeMode);
+    }

-    const MaximizeMode oldMode = m_requestedMaximizeMode;
-    const QRectF oldGeometry = moveResizeGeometry();
+    MaximizeMode currentMaximizeMode = m_requestedMaximizeMode;

-    mode = rules()->checkMaximize(mode);
-    if (m_requestedMaximizeMode == mode) {
+    if (currentMaximizeMode == nextMaximizeMode) {
         return;
     }

-    Q_EMIT maximizedAboutToChange(mode);
-    m_requestedMaximizeMode = mode;
+    // backup the next maximize mode (it will become the currentMaximizeMode in the next round)
+    m_requestedMaximizeMode = nextMaximizeMode;

-    // call into decoration update borders
-    if (m_nextDecoration && !(options->borderlessMaximizedWindows() && m_requestedMaximizeMode == MaximizeFull)) {
-        changeMaximizeRecursion = true;
-        const auto c = m_nextDecoration->window();
-        if ((m_requestedMaximizeMode & MaximizeVertical) != (oldMode & MaximizeVertical)) {
-            Q_EMIT c->maximizedVerticallyChanged(m_requestedMaximizeMode & MaximizeVertical);
-        }
-        if ((m_requestedMaximizeMode & MaximizeHorizontal) != (oldMode & MaximizeHorizontal)) {
-            Q_EMIT c->maximizedHorizontallyChanged(m_requestedMaximizeMode & MaximizeHorizontal);
-        }
-        if ((m_requestedMaximizeMode == MaximizeFull) != (oldMode == MaximizeFull)) {
-            Q_EMIT c->maximizedChanged(m_requestedMaximizeMode == MaximizeFull);
-        }
-        changeMaximizeRecursion = false;
-    }
+    animateMaximize(nextMaximizeMode, currentMaximizeMode);

-    if (options->borderlessMaximizedWindows()) {
-        setNoBorder(m_requestedMaximizeMode == MaximizeFull);
+    if (nextMaximizeMode != MaximizeShade && options->borderlessMaximizedWindows()) {
+        setNoBorder(nextMaximizeMode == MaximizeFull);
     }

-    if (!restore.isNull()) {
+    if (restore.isNull()) {
+        backupCurrentMaximizeGeometry(currentMaximizeMode);
+    } else {
         setGeometryRestore(restore);
-    } else {
-        if (requestedQuickTileMode() == QuickTileMode(QuickTileFlag::None)) {
-            QRectF savedGeometry = geometryRestore();
-            if (!(oldMode & MaximizeVertical)) {
-                savedGeometry.setTop(oldGeometry.top());
-                savedGeometry.setBottom(oldGeometry.bottom());
-            }
-            if (!(oldMode & MaximizeHorizontal)) {
-                savedGeometry.setLeft(oldGeometry.left());
-                savedGeometry.setRight(oldGeometry.right());
-            }
-            setGeometryRestore(savedGeometry);
-        }
     }

-    if (m_requestedMaximizeMode != MaximizeRestore) {
+    if (nextMaximizeMode != MaximizeRestore) {
         exitQuickTileMode();
     }

-    QRectF geometry = oldGeometry;
+    const QRectF clientArea = isElectricBorderMaximizing() ? workspace()->clientArea(MaximizeArea, this, interactiveMoveResizeAnchor()) : workspace()->clientArea(MaximizeArea, this, moveResizeOutput());
+    QRectF maximizeHorizontalGeometry = nextMaximizeHorizontalGeometry(clientArea, nextMaximizeMode);
+    QRectF maximizeVerticalGeometry = nextMaximizeVerticalGeometry(clientArea, nextMaximizeMode);

-    if (m_requestedMaximizeMode & MaximizeHorizontal) {
-        // Stretch the window vertically to fit the size of the maximize area.
-        geometry.setX(clientArea.x());
-        geometry.setWidth(clientArea.width());
-    } else if (oldMode & MaximizeHorizontal) {
-        if (geometryRestore().isValid()) {
-            // The window is no longer maximized horizontally and the saved geometry is valid.
-            geometry.setX(geometryRestore().x());
-            geometry.setWidth(geometryRestore().width());
-        } else {
-            // The window is no longer maximized horizontally and the saved geometry is
-            // invalid. This would happen if the window had been mapped in the maximized state.
-            // We ask the client to resize the window horizontally to its preferred size.
-            geometry.setX(clientArea.x());
-            geometry.setWidth(0);
-        }
-    }
+    QRectF nextMaximizeGeometry = moveResizeGeometry();
+    nextMaximizeGeometry.setX(maximizeHorizontalGeometry.x());
+    nextMaximizeGeometry.setWidth(maximizeHorizontalGeometry.width());
+    nextMaximizeGeometry.setY(maximizeVerticalGeometry.y());
+    nextMaximizeGeometry.setHeight(maximizeVerticalGeometry.height());

-    if (m_requestedMaximizeMode & MaximizeVertical) {
-        // Stretch the window horizontally to fit the size of the maximize area.
-        geometry.setY(clientArea.y());
-        geometry.setHeight(clientArea.height());
-    } else if (oldMode & MaximizeVertical) {
-        if (geometryRestore().isValid()) {
-            // The window is no longer maximized vertically and the saved geometry is valid.
-            geometry.setY(geometryRestore().y());
-            geometry.setHeight(geometryRestore().height());
-        } else {
-            // The window is no longer maximized vertically and the saved geometry is
-            // invalid. This would happen if the window had been mapped in the maximized state.
-            // We ask the client to resize the window vertically to its preferred size.
-            geometry.setY(clientArea.y());
-            geometry.setHeight(0);
-        }
-    }
-
-    moveResize(geometry);
+    // emit signal to resize window
+    moveResize(nextMaximizeGeometry);
     markAsPlaced();

-    doSetMaximized();
+    doSetMaximized(nextMaximizeMode, currentMaximizeMode);
 }

 XdgPopupWindow::XdgPopupWindow(XdgPopupInterface *shellSurface)
Index: src/xdgshellwindow.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/xdgshellwindow.h b/src/xdgshellwindow.h
--- a/src/xdgshellwindow.h	(revision 9739e3174abaaa97a3008f6933af4642bc26b800)
+++ b/src/xdgshellwindow.h	(revision 467bd2c1d82f46d0b2629411ef9dbec1f7d01f57)
@@ -179,7 +179,7 @@
     void doMinimize() override;
     void doSetActive() override;
     void doSetFullScreen();
-    void doSetMaximized();
+    void doSetMaximized(MaximizeMode nextMaximizeMode, MaximizeMode currentMaximizeMode);
     bool doStartInteractiveMoveResize() override;
     void doFinishInteractiveMoveResize() override;
     bool acceptsFocus() const override;
@@ -190,6 +190,8 @@
     void doSetPreferredColorDescription() override;

 private:
+    void setMaximizeModeToStates(MaximizeMode nextMaximizeMode);
+    MaximizeMode getMaximizeModeFromStates(const XdgToplevelInterface::States &states);
     void handleWindowTitleChanged();
     void handleAppIdChanged();
     void handleWindowMenuRequested(SeatInterface *seat,
@@ -211,6 +213,10 @@
     void handleMinimumSizeChanged();
     void initialize();
     void updateMaximizeMode(MaximizeMode maximizeMode);
+    void animateMaximize(MaximizeMode nextMaximizeMode, MaximizeMode currentMaximizeMode);
+    QRectF nextMaximizeHorizontalGeometry(const QRectF &clientArea, MaximizeMode nextMaximizeMode);
+    QRectF nextMaximizeVerticalGeometry(const QRectF &clientArea, MaximizeMode nextMaximizeMode);
+    void backupCurrentMaximizeGeometry(MaximizeMode currentMaximizeMode);
     void updateFullScreenMode(bool set);
     void sendPing(PingReason reason);
     QPointF initialPosition() const;
