Subject: [PATCH] jacks-customizations: disable natural swiping
---
Index: src/virtualdesktops.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/virtualdesktops.cpp b/src/virtualdesktops.cpp
--- a/src/virtualdesktops.cpp	(revision b73960c196fcd9a5cefc56499eff51f07afd3af6)
+++ b/src/virtualdesktops.cpp	(revision 1a24f02af8383f470c1cd4aa676c11d5d848f271)
@@ -771,8 +771,8 @@
     };
     // input()->registerTouchpadSwipeShortcut(SwipeDirection::Left, 3, m_swipeGestureReleasedX.get(), left);
     // input()->registerTouchpadSwipeShortcut(SwipeDirection::Right, 3, m_swipeGestureReleasedX.get(), right);
-    input()->registerTouchpadSwipeShortcut(SwipeDirection::Left, 4, m_swipeGestureReleasedX.get(), left);
-    input()->registerTouchpadSwipeShortcut(SwipeDirection::Right, 4, m_swipeGestureReleasedX.get(), right);
+    input()->registerTouchpadSwipeShortcut(SwipeDirection::Left, 4, m_swipeGestureReleasedX.get(), right);
+    input()->registerTouchpadSwipeShortcut(SwipeDirection::Right, 4, m_swipeGestureReleasedX.get(), left);
     // input()->registerTouchpadSwipeShortcut(SwipeDirection::Down, 3, m_swipeGestureReleasedY.get(), [this](qreal cb) {
     //     if (grid().height() > 1) {
     //         m_currentDesktopOffset.setY(-cb);
