From: Jacks Diao <kdha200501@gmail.com>
Subject: [PATCH] jacks-customizations: interpret the enter QKeyEvent on file as rename signal
---
Index: src/kitemviews/kitemlistcontroller.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/kitemviews/kitemlistcontroller.cpp b/src/kitemviews/kitemlistcontroller.cpp
--- a/src/kitemviews/kitemlistcontroller.cpp	(revision 6b46cf33ebdf13cc048caa1d679412af57d31410)
+++ b/src/kitemviews/kitemlistcontroller.cpp	(revision 1bf1debd749c964f54fc80b9a3af9bfcdd311ab6)
@@ -8,6 +8,7 @@
  */

 #include "kitemlistcontroller.h"
+#include "views/dolphinview.h"

 #include "kitemlistselectionmanager.h"
 #include "kitemlistview.h"
@@ -32,6 +33,7 @@

 KItemListController::KItemListController(KItemModelBase *model, KItemListView *view, QObject *parent)
     : QObject(parent)
+    , m_parent(parent)
     , m_singleClickActivationEnforced(false)
     , m_selectionMode(false)
     , m_selectionTogglePressed(false)
@@ -449,14 +451,18 @@

     case Qt::Key_Enter:
     case Qt::Key_Return: {
-        const KItemSet selectedItems = m_selectionManager->selectedItems();
-        if (selectedItems.count() >= 2) {
-            Q_EMIT itemsActivated(selectedItems);
-        } else if (selectedItems.count() == 1) {
-            Q_EMIT itemActivated(selectedItems.first());
-        } else {
-            Q_EMIT itemActivated(index);
-        }
+        // const KItemSet selectedItems = m_selectionManager->selectedItems();
+        // if (selectedItems.count() >= 2) {
+        //     Q_EMIT itemsActivated(selectedItems);
+        // } else if (selectedItems.count() == 1) {
+        //     Q_EMIT itemActivated(selectedItems.first());
+        // } else {
+        //     Q_EMIT itemActivated(index);
+        // }
+
+        // When the KStandardItemListWidget is in edit mode, subsequent Enter/Return QKeyEvent are not captured here
+        DolphinView *dolphinView = qobject_cast<DolphinView *>(m_parent);
+        dolphinView->renameSelectedItems();
         break;
     }

Index: src/kitemviews/kitemlistcontroller.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/kitemviews/kitemlistcontroller.h b/src/kitemviews/kitemlistcontroller.h
--- a/src/kitemviews/kitemlistcontroller.h	(revision 6b46cf33ebdf13cc048caa1d679412af57d31410)
+++ b/src/kitemviews/kitemlistcontroller.h	(revision 1bf1debd749c964f54fc80b9a3af9bfcdd311ab6)
@@ -342,6 +342,7 @@
     void startRubberBand();

 private:
+    QObject *m_parent;
     bool m_singleClickActivationEnforced;
     bool m_selectionMode;
     bool m_selectionTogglePressed;
