Subject: [PATCH] jacks-customizations: fix desktop icon label text
---
Index: containments/desktop/package/contents/ui/FolderItemDelegate.qml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/containments/desktop/package/contents/ui/FolderItemDelegate.qml b/containments/desktop/package/contents/ui/FolderItemDelegate.qml
--- a/containments/desktop/package/contents/ui/FolderItemDelegate.qml	(revision 3230ecdae8c8198d2bc7644942475c25108615e3)
+++ b/containments/desktop/package/contents/ui/FolderItemDelegate.qml	(revision b62854fe5d29b0ce042f78ae204bf7884be3903d)
@@ -373,7 +373,7 @@
 
                     text: main.nameWrapped
                     font.italic: model.isLink
-                    wrapMode: (maximumLineCount === 1) ? Text.NoWrap : Text.Wrap
+                    wrapMode: Text.NoWrap
                     horizontalAlignment: Text.AlignHCenter
                 }
 
Index: containments/desktop/package/contents/ui/RenameEditor.qml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/containments/desktop/package/contents/ui/RenameEditor.qml b/containments/desktop/package/contents/ui/RenameEditor.qml
--- a/containments/desktop/package/contents/ui/RenameEditor.qml	(revision 3230ecdae8c8198d2bc7644942475c25108615e3)
+++ b/containments/desktop/package/contents/ui/RenameEditor.qml	(revision b62854fe5d29b0ce042f78ae204bf7884be3903d)
@@ -43,6 +43,8 @@
 
         rightPadding: root.PlasmaComponents.ScrollBar.vertical.visible ? root.PlasmaComponents.ScrollBar.vertical.width : 0
 
+        color: black
+
         Kirigami.SpellCheck.enabled: false
 
         property Item targetItem: null
@@ -198,4 +200,3 @@
         }
     }
 }
-
