From: Jacks Diao <kdha200501@gmail.com>
Subject: [PATCH] jacks-customizations: do not include background when caching folder item image for dragging
---
Index: containments/desktop/package/contents/ui/FolderItemDelegate.qml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/containments/desktop/package/contents/ui/FolderItemDelegate.qml b/containments/desktop/package/contents/ui/FolderItemDelegate.qml
--- a/containments/desktop/package/contents/ui/FolderItemDelegate.qml	(revision 058d4110ec9d06ae95a33f9a5d6a9e89254d723e)
+++ b/containments/desktop/package/contents/ui/FolderItemDelegate.qml	(revision 67a29dfd0a0d9d54bff16340a00a0d39bfd64a65)
@@ -37,6 +37,7 @@
     property Item hoverArea:       loader.item ? loader.item.hoverArea      : null
     property Item frame:           loader.item ? loader.item.frame          : null
     property Item toolTip:         loader.item ? loader.item.toolTip        : null
+    property bool takingSnapshot: false;
     Accessible.name: name
     Accessible.role: Accessible.Canvas
 
@@ -82,8 +83,10 @@
 
     function updateDragImage() {
         if (selected && !blank) {
+            takingSnapshot = true;
             loader.grabToImage(result => {
                 dir.addItemDragImage(positioner.map(index), main.x + loader.x, main.y + loader.y, loader.width, loader.height, result.image);
+                takingSnapshot = false;
             });
         }
     }
@@ -232,7 +235,7 @@
                 property string prefix: ""
 
                 sourceComponent: frameComponent
-                active: impl.iconAndLabelsShouldlookSelected || model.selected
+                active: takingSnapshot ? false : impl.iconAndLabelsShouldlookSelected || model.selected
                 asynchronous: true
 
                 width: {
@@ -287,7 +290,7 @@
                     height: main.GridView.view.iconSize
 
                     opacity: {
-                        if (root.useListViewMode && selectionButton.visible) {
+                        if (root.useListViewMode && selectionButton || takingSnapshot) {
                             return 0.3;
                         }
 
